File: app/floofycrew/floofy_loader/events.py
"""The Loader event bus (Requirement 3.8) — consumable by mods (``ctx.events`` / ``floofy.events``) and the manager UI.

Names the Loader publishes (:data:`EVENTS`):

* ``mod.activated`` / ``mod.deactivated`` — ``{"mod", "version"}``;
* ``mod.quarantined`` — ``{"mod", "version", "hostVersion"}`` (a ``broken`` matrix cell);
* ``mod.faulted`` — ``{"mod", "source", "message"}`` (hook fault or SPA fault report);
* ``host.version_changed`` — ``{"previous", "current"}`` (the last recorded boot ran another host version);
* ``loader.state`` — ``{"loader", "active", "bootedAt"}`` after every boot and reload.

Delivery is synchronous and fail-open per subscriber: a raising subscriber is
logged with attribution and counted, the others still receive the event. Mods
may publish their own names (``mod.<id>.<topic>`` by convention); every
delivered event is kept in a short history the state API exposes.
"""
from __future__ import annotations

import logging
from collections import deque
from dataclasses import dataclass, field
from typing import Any, Callable

from .state import utc_now

__all__ = ["EVENTS", "EventBus", "ModEvents", "Subscription"]

logger = logging.getLogger("floofy.loader.events")

EVENTS = ("mod.activated", "mod.deactivated", "mod.quarantined", "mod.faulted", "host.version_changed", "loader.state")
WILDCARD = "*"
HISTORY = 200

Handler = Callable[[str, dict[str, Any]], Any]


@dataclass(frozen=True)
class Subscription:
    name: str
    fn: Handler
    mod_id: str | None
    sequence: int

    def to_dict(self) -> dict[str, Any]:
        return {"event": self.name, "mod": self.mod_id, "fn": getattr(self.fn, "__qualname__", repr(self.fn))}


@dataclass
class EventBus:
    """Synchronous pub/sub with per-mod attribution."""

    subscriptions: list[Subscription] = field(default_factory=list)
    history: deque = field(default_factory=lambda: deque(maxlen=HISTORY))
    errors: list[dict[str, str]] = field(default_factory=list)
    _sequence: int = 0

    def subscribe(self, name: str, fn: Handler, mod_id: str | None = None) -> Subscription:
        if not callable(fn):
            raise TypeError("event handler must be callable")
        self._sequence += 1
        subscription = Subscription(name, fn, mod_id, self._sequence)
        self.subscriptions.append(subscription)
        return subscription

    def unsubscribe(self, subscription: Subscription) -> bool:
        if subscription in self.subscriptions:
            self.subscriptions.remove(subscription)
            return True
        return False

    def unsubscribe_mod(self, mod_id: str) -> int:
        mine = [s for s in self.subscriptions if s.mod_id == mod_id]
        for subscription in mine:
            self.subscriptions.remove(subscription)
        return len(mine)

    def publish(self, name: str, payload: dict[str, Any] | None = None, *, source: str = "loader") -> int:
        """Deliver ``name`` to every matching subscriber, fail-open; returns the number delivered."""
        data = dict(payload or {})
        record = {"event": name, "payload": data, "source": source, "ts": utc_now()}
        self.history.append(record)
        delivered = 0
        for subscription in list(self.subscriptions):
            if subscription.name != name and subscription.name != WILDCARD:
                continue
            try:
                subscription.fn(name, data)
                delivered += 1
            except Exception as exc:  # noqa: BLE001 - one bad subscriber never blocks the others
                who = subscription.mod_id or "loader"
                logger.exception("event %s: subscriber of %s raised %s", name, who, exc)
                self.errors.append({"event": name, "mod": who, "error": f"{type(exc).__name__}: {exc}"})
        return delivered

    def to_dict(self) -> dict[str, Any]:
        return {
            "subscriptions": [s.to_dict() for s in self.subscriptions],
            "history": list(self.history)[-20:],
            "errors": list(self.errors)[-20:],
            "events": list(EVENTS),
        }


@dataclass
class ModEvents:
    """``ctx.events`` — the bus bound to one mod (subscriptions carry its id and are dropped with it)."""

    bus: EventBus
    mod_id: str

    def subscribe(self, name: str, fn: Handler) -> Subscription:
        return self.bus.subscribe(name, fn, self.mod_id)

    def unsubscribe(self, subscription: Subscription) -> bool:
        return self.bus.unsubscribe(subscription)

    def publish(self, name: str, payload: dict[str, Any] | None = None) -> int:
        return self.bus.publish(name, payload, source=f"mod:{self.mod_id}")

    @property
    def history(self) -> list[dict[str, Any]]:
        return list(self.bus.history)

File: app/floofycrew/ui/events.mjs
// FloofyCrew SPA host — the dashboard-side event bus (Requirement 3.8, 4.1).
//
// A tiny fail-open pub/sub: a subscriber that throws is logged and counted,
// the other subscribers still receive the event, and the emitter never sees
// the exception. `scoped(bus, modId)` gives a mod a view that remembers its
// subscriptions so the host can drop them all when the mod faults.
//
// Event names the host publishes: `mod.loaded`, `mod.faulted`,
// `surfaces.changed`, `route.changed`, `host.ready`. `*` subscribes to
// everything (the handler receives the event name as its first argument).
"use strict";

/**
 * @typedef {(name: string, payload: unknown) => void} Handler
 */

/**
 * Create a new bus.
 * @param {{log?: (message: string, error?: unknown) => void, historyLimit?: number}} [options]
 */
export function createBus(options = {}) {
  const log = options.log ?? ((message, error) => console.warn(`[floofy] ${message}`, error ?? ""));
  const historyLimit = options.historyLimit ?? 50;
  /** @type {Map<string, Set<{handler: Handler, owner: string|null}>>} */
  const subscribers = new Map();
  /** @type {{name: string, payload: unknown, at: number}[]} */
  const history = [];
  let errors = 0;

  function bucket(name) {
    let set = subscribers.get(name);
    if (!set) {
      set = new Set();
      subscribers.set(name, set);
    }
    return set;
  }

  /**
   * Subscribe `handler` to `name` (`*` for every event). Returns an unsubscribe function.
   * @param {string} name
   * @param {Handler} handler
   * @param {string|null} [owner] mod id used for bulk removal
   */
  function on(name, handler, owner = null) {
    if (typeof handler !== "function") throw new TypeError("handler must be a function");
    const entry = { handler, owner };
    bucket(name).add(entry);
    return () => {
      bucket(name).delete(entry);
    };
  }

  /** Subscribe for exactly one delivery. */
  function once(name, handler, owner = null) {
    const off = on(
      name,
      (eventName, payload) => {
        off();
        handler(eventName, payload);
      },
      owner,
    );
    return off;
  }

  /** Remove every subscription of `handler` on `name`. */
  function off(name, handler) {
    for (const entry of bucket(name)) {
      if (entry.handler === handler) bucket(name).delete(entry);
    }
  }

  /**
   * Deliver `payload` to the subscribers of `name` and of `*`, synchronously and fail-open.
   * @returns {number} how many handlers ran without throwing
   */
  function emit(name, payload) {
    history.push({ name, payload, at: Date.now() });
    if (history.length > historyLimit) history.shift();
    let delivered = 0;
    for (const set of [subscribers.get(name), subscribers.get("*")]) {
      if (!set) continue;
      for (const entry of [...set]) {
        try {
          entry.handler(name, payload);
          delivered += 1;
        } catch (error) {
          errors += 1;
          log(`event ${name}: subscriber${entry.owner ? ` of ${entry.owner}` : ""} threw`, error);
        }
      }
    }
    return delivered;
  }

  /** Drop every subscription registered with `owner`. Returns how many were dropped. */
  function dropOwner(owner) {
    let dropped = 0;
    for (const set of subscribers.values()) {
      for (const entry of [...set]) {
        if (entry.owner === owner) {
          set.delete(entry);
          dropped += 1;
        }
      }
    }
    return dropped;
  }

  /** Snapshot for `window.floofy.events.inspect()` and the reporter. */
  function inspect() {
    const subscriptions = {};
    for (const [name, set] of subscribers) {
      if (set.size) subscriptions[name] = [...set].map((entry) => entry.owner ?? "host");
    }
    return { subscriptions, history: history.slice(-20), errors };
  }

  return Object.freeze({ on, once, off, emit, dropOwner, inspect });
}

/**
 * A mod-scoped view of `bus`: same `on/once/off/emit`, every subscription attributed to `modId`.
 * @param {ReturnType<typeof createBus>} bus
 * @param {string} modId
 */
export function scoped(bus, modId) {
  return Object.freeze({
    on: (name, handler) => bus.on(name, handler, modId),
    once: (name, handler) => bus.once(name, handler, modId),
    off: (name, handler) => bus.off(name, handler),
    emit: (name, payload) => bus.emit(name, { ...(payload && typeof payload === "object" ? payload : { value: payload }), source: modId }),
    inspect: () => bus.inspect(),
  });
}
